keps/sig-node/2535-ensure-secret-pulled-images/README.md

```diff
@@ -0,0 +1,379 @@
+the `pull` does not re-pull already-pulled layers of the image, but simply
+resolves/verifies the image manifest has not changed in the registry (which
+incidentally requires authenticating to private registries, which enforces the
+image access). That means in the normal case (where the image has not changed
+since the last pull), the request size is O(kb). However, the `pull` does put
+the registry in the critical path of starting a container, since an unavailable
+registry will fail the pull image manifest check (with or without proper
+authentication.)
+
+### Goals
+
+Modify the current pullIfNotPresent policy management enforced by `kubelet` to
+ensure the images pulled with a secret by `kublet` since boot. During the
+EnsureImagesExist step `kubelet` will require authentication of present images
+pulled with auth since boot.
+
+Optimize to only force re-authentication for a pod container image when the
+secret used to pull the container image is not present. IOW if an image is
+pulled with authentication for a first pod, subsequent pods that have the same
+authentication information should not need to re-authenticate.
+
+Images already present at boot or loaded externally to `kubelet` or successfully
+pulled through `kubelet` with no imagePullSecret/authentication required will
+not require authentication.
+
+### Non-Goals
+
+Out of scope for this KEP is an image caching policy that would direct container
+runtimes through the CRI wrt. how they should treat the caching of images on a
+node. Such as store for public use but only if encrypted. Or Store for private
+use un-encrypted...
+
+## Proposal
+
+`kubelet` will keep a list, since boot, of container images that required
+authentication and a list of the authentications that successfully pulled the image.
+
+`kubelet` will ensure any image in the list is always pulled if an authentication
+used is not present, thus enforcing authentication / re-authentication.
+
+
+### User Stories
+
+#### Story 1
+User with multiple tenants will be able to support all image pull policies without
+concern that one tenant will gain access to an image that they don't have rights to.
+
+#### Story 2
+User will will no longer have to inject the Pull Always Image Pull Policy to
+ensure all tenants have rights to the images that are already present on a host.
+
+
+### Notes/Constraints/Caveats (Optional)
+
+With the default of the feature gate being off, users / cloud providers will have
+to set the feature gate to true to gain these this Secure by Default benefit.
+
+### Risks and Mitigations
+
+Image authentications with a registry may expire. To mitigate expirations a
+a timeout could be used to force re-authentication. The timeout could be a
+container runtime feature or a `kubelet` feature. If at the container runtime,
+images would not be present during the EnsureImagesExist step, thus would have
+to be pulled and authenticated if necessary.
```

Comment:
Maybe also include re-auth on boot?

Reply:
yes, an optional re-auth on boot switch would go well with the optional timeout adding it to the mitigation ideas..